File: Jan2021/03Jan/BeautifulArrangement.cpp
/*
Problem name: 526. Beautiful Arrangement
Problem link: https://leetcode.com/problems/beautiful-arrangement/
*/

class Solution {
public:
    int ans=0;
    int countArrangement(int n) {
        vector<bool> visited(n);
        helper(visited, n);
        return ans;
    }
    void helper(vector<bool>& visited, int end, int pos = 1){
        if(pos == end+1){
            ans++;
            return;
        }
        for(int i=1; i<=end; ++i){
            if(!visited[i] && (i%pos==0 || pos%i==0)){
                visited[i]=true;
                helper(visited, end, pos+1 );
                visited[i]=false;
            }
        }
    }
};

File: Jan2021/03Jan/BeautifulArrangement.java
/*
Problem Name: 526. Beautiful Arrangement
Problem Link: https://leetcode.com/problems/beautiful-arrangement/
*/
class Solution {
    int arrangements;
    public int countArrangement(int n) {
        boolean visited[] = new boolean[n+1];
        arrangements = 0;
        generate(1, visited);
        return arrangements;
    }
    void generate(int curInd, boolean v[]){
        if(curInd == v.length){
            arrangements++;
            return;
        }
        int n = v.length;
        for(int i=1;i<n;i++){
            if(!v[i] && ((i%curInd == 0) || (curInd%i == 0))){
                v[i] = true;
                generate(curInd + 1, v);
                v[i] = false;
            }
        }
    }
}
